### PR TITLE
feat: scene runtime factory

### PR DESCRIPTION
**Pull request by dalkia**
- Creates a Factory for SceneRuntime. It handles the cahcing of reusable javascript files
- Moves local files to Streaming Assets folder
- Retrieves the files using WebRequest. Its set up so we can pass a web url and the code should be retrieve accordingly. For now, this is the only code content provider
- Deletes the use of the Helper class is the SceneRuntime constructor and delegates all its responsibility to the Factory
- I decided not to bring the whole AssetResolver base logic from the old renderer yet, as it was a total overkill for this requirement. Most probably we will need it when we start downloading assets, but for now it seemed unnecesary.

### CODE REVIEW

File: Explorer/Assets/StreamingAssets/Scenes/Cube/cube.js
Comment (kuruk-mm) [inline Explorer/Assets/StreamingAssets/Scenes/Cube/cube.js:1]: those scenes are just for testing, we must not put them in the build

Comment (dalkia) [reply Explorer/Assets/StreamingAssets/Scenes/Cube/cube.js:1]: Moved it outside of the project into TestResources

Comment (mikhail-dcl) [reply Explorer/Assets/StreamingAssets/Scenes/Cube/cube.js:1]: What are we going to test with? Please keep in mind that at certain stage when the MVP systems are integrated we will have to try on the build as well

**Review by mikhail-dcl — COMMENTED** [review]
I have just minor comments

File: Explorer/Assets/Scripts/AssetManagement/CodeResolver/ICodeContentProvider.cs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+using Cysharp.Threading.Tasks;
+using System.Threading;
+
+namespace AssetManagement.CodeResolver
+{
+    public interface ICodeContentProvider
```
[the remark this answers is not in the document]
Comment (dalkia) [reply Explorer/Assets/Scripts/AssetManagement/CodeResolver/ICodeContentProvider.cs:6]: Done!

File: Explorer/Assets/Scripts/AssetManagement/CodeResolver/WebCodeContentProvider.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        async UniTask<string> RequestFile(string path, CancellationToken cancellationToken)
+        {
+            var request = UnityWebRequest.Get(path);
+
+            await request.SendWebRequest().WithCancellation(cancellationToken);
+
+            if (request.result == UnityWebRequest.Result.Success)
+                return request.downloadHandler.text;
+
+            throw new Exception($"Asset request failed with error {request.error}");
```
[the remark this answers is not in the document]
Comment (dalkia) [reply Explorer/Assets/Scripts/AssetManagement/CodeResolver/WebCodeContentProvider.cs:22]: Im using this throw Exception not only for the Await exception, but for a request fail. Maybe adding an else makes it more clear?

File: Explorer/Assets/Scripts/AssetManagement/CodeResolver/WebCodeContentProvider.cs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+using UnityEngine.Networking;
+
+namespace AssetManagement.CodeResolver
+{
+    public class WebCodeContentProvider : ICodeContentProvider
+    {
+        public UniTask<string> GetCodeAsync(string url, CancellationToken cancellationToken = default) =>
+            RequestFile(url, cancellationToken);
+
+        async UniTask<string> RequestFile(string path, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (dalkia) [reply Explorer/Assets/Scripts/AssetManagement/CodeResolver/WebCodeContentProvider.cs:13]: I had the same doubt myself. The thing is that `EMBEDDED` will still make a web request to a file, so its gonna be basically the same code as the `WEB` provider. Maybe `EMBEDDED` would be something we load from Resources, which in this case we dont have. WDYT?

File: Explorer/Assets/Scripts/AssetManagement/CodeResolver/WebCodeContentProvider.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikhail-dcl) [reply Explorer/Assets/Scripts/AssetManagement/CodeResolver/WebCodeContentProvider.cs]: Take a look at this
```
public struct UnityWebRequestAsyncOperationAwaiter : ICriticalNotifyCompletion
        {
            UnityWebRequestAsyncOperation asyncOperation;
            Action<AsyncOperation> continuationAction;

            public UnityWebRequestAsyncOperationAwaiter(UnityWebRequestAsyncOperation asyncOperation)
            {
                this.asyncOperation = asyncOperation;
                this.continuationAction = null;
            }

            public bool IsCompleted => asyncOperation.isDone;

            public UnityWebRequest GetResult()
            {
                if (continuationAction != null)
                {
                    asyncOperation.completed -= continuationAction;
                    continuationAction = null;
                    var result = asyncOperation.webRequest;
                    asyncOperation = null;
                    if (result.IsError())
                    {
                        throw new UnityWebRequestException(result);
                    }
                    return result;
                }
                else
                {
                    var result = asyncOperation.webRequest;
                    asyncOperation = null;
                    if (result.IsError())
                    {
                        throw new UnityWebRequestException(result);
                    }
                    return result;
                }
            }

            public void OnCompleted(Action continuation)
            {
                UnsafeOnCompleted(continuation);
            }

            public void UnsafeOnCompleted(Action continuation)
            {
                Error.ThrowWhenContinuationIsAlreadyRegistered(continuationAction);
                continuationAction = PooledDelegate<AsyncOperation>.Create(continuation);
                asyncOperation.completed += continuationAction;
            }
        }
```


```
if (result.IsError())
                    {
                        throw new UnityWebRequestException(result);
                    }
```

it already throws an exception if the result is an error
```
public static bool IsError(this UnityWebRequest unityWebRequest)
        {
#if UNITY_2020_2_OR_NEWER
            var result = unityWebRequest.result;
            return (result == UnityWebRequest.Result.ConnectionError)
                || (result == UnityWebRequest.Result.DataProcessingError)
                || (result == UnityWebRequest.Result.ProtocolError);
#else
            return unityWebRequest.isHttpError || unityWebRequest.isNetworkError;
#endif
        }
```

[the remark this answers is not in the document]
Comment (dalkia) [reply Explorer/Assets/Scripts/AssetManagement/CodeResolver/WebCodeContentProvider.cs]: Aha, got it. Yeah, you are completely right. That line is redudant

**Review by mikhail-dcl — APPROVED** [review]

**Pull request merged**